scheduler/src/cook/mesos/sandbox.clj

```diff
@@ -182,11 +189,64 @@
 
 (defn sync-agent-sandboxes
   "Asynchronously triggers state syncing from the mesos agent."
-  [{:keys [mesos-agent-query-cache task-id->sandbox-agent]} framework-id agent-hostname]
+  [{:keys [mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent]} framework-id agent-hostname]
   (when (and mesos-agent-query-cache task-id->sandbox-agent framework-id agent-hostname)
     (future
       (refresh-agent-cache-entry
-        framework-id mesos-agent-query-cache task-id->sandbox-agent agent-hostname))))
+        framework-id mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent agent-hostname))))
+
+(defn start-sandbox-publisher
+  "Launches a timer task that triggers publishing of the task-id->sandbox state to datomic.
+   The task is invoked at intervals of publish-interval-ms ms."
+  [task-id->sandbox-agent datomic-conn publish-batch-size publish-interval-ms]
+  (log/info "Starting sandbox publisher at intervals of" publish-interval-ms "ms")
+  (chime/chime-at
+    (periodic/periodic-seq (time/now) (time/millis publish-interval-ms))
+    (fn sandbox-publisher-task [_]
+      (log/info "Requesting publishing of instance sandbox directories")
+      (publish-sandbox-to-datomic! datomic-conn publish-batch-size task-id->sandbox-agent))
+    {:error-handler (fn sandbox-publisher-error-handler [ex]
+                      (log/error ex "Instance sandbox directory publish failed"))}))
+
+(defn start-host-sandbox-syncer
+  "Launches a timer task that triggers syncing of sandbox directories of any pending hosts."
+  [mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent sync-interval-ms max-consecutive-sync-failure]
+  (log/info "Starting sandbox syncer at intervals of" sync-interval-ms "ms")
+  (let [syncer-state {:mesos-agent-query-cache mesos-agent-query-cache
+                      :pending-sync-agent pending-sync-agent
+                      :task-id->sandbox-agent task-id->sandbox-agent}]
+    (chime/chime-at
+      (periodic/periodic-seq (time/now) (time/millis sync-interval-ms))
+      (fn host-sandbox-syncer-task [_]
+        (let [{:keys [framework-id host->consecutive-failures pending-sync-hosts]} @pending-sync-agent
+              num-pending-sync-hosts (count pending-sync-hosts)]
+          (log/info num-pending-sync-hosts "hosts have pending syncs")
+          (loop [[hostname & remaining-hostnames] (vec pending-sync-hosts)
+                 pending-sync-agent-send-performed false]
+            (let [consecutive-failures (get host->consecutive-failures hostname 0)]
+              (cond
+                (nil? hostname)
+                (when pending-sync-agent-send-performed
```

Comment:
Would it be that bad to just have the `await` regardless of whether we sent anything? I don't think it's costing us too much, and it would be nice to clean up the logic a bit.

Reply:
I'd rather only trigger the wait if this iteration performed any agent operations.

Reply:
👍